bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/SortedLedgerStorage.java

```diff
@@ -209,16 +211,39 @@ public void onSizeLimitReached(final Checkpoint cp) throws IOException {
             public void run() {
                 try {
                     LOG.info("Started flushing mem table.");
-                    long logIdBeforeFlush = entryLogger.getCurrentLogId();
+                    long prevAllocLogIdBeforeFlush = entryLogger.getPreviousAllocatedEntryLogId();
```

Comment:
This change is not correct. Because it should be "current" log id, not "previous allocated entry log id". That says the behavior is different between preallocation enabled and disabled.

That leads me to think that current entry log manager interface is not abstracted with right methods. because it abstracts a lot of methods around the underlying files, which underlying files are kind of tight with corresponding checkpointing and flushing logic. two different entry log manager can't really fit in this interface or checkpointing logic as well. 

So as what I said in my previous comment, we need to think about what is the right methods to put in EntryLogManager.

> I would suggest the EntryLogManager should hide the details on how to create log, allocate log,
> flush/checkpoint logs. the details such as preallocationLogId, unflushedLogId and how to lock them should be > hidden to the implementation, rather than defining a too finer interface here.

Reply:
"This change is not correct. Because it should be "current" log id, not "previous allocated entry log id". That says the behavior is different between preallocation enabled and disabled."

@sijie why would you say that behavior would be different between preallocation enabled and disabled? Is it because of async nature of preallocation?

Ok, I think I can do the following, which would solve 2 problems here.

I can add "getCurrentLogId" to the EntryLogManager Interface. For EntryLogManagerForSingleEntryLog implementation it would give the logId of the current activelog (which is current behavior). For EntryLogManagerForEntryLogPerLedger it would give some constant value all the times and it should be fine in entryLogPerLedger case, because checkpoint is not made when entrylog is rotated but instead SyncThread drives it periodically. 

Also it would remove the need of extra condition regarding "isTransactionalCompactionEnabled" which I added in this change.

So it would be clean and straightforward.

Reply:
@reddycharan 

>  why would you say that behavior would be different between preallocation enabled and disabled? Is it because of async nature of preallocation?

yeah because preallocation is in background, preallocation can advance entry log id. if it does, the logic is not correct.

> I can add "getCurrentLogId" to the EntryLogManager Interface. For EntryLogManagerForSingleEntryLog implementation it would give the logId of the current activelog (which is current behavior).

That could work for now. Although I would still like to see if there is possibility to produce a cleaner interface without exposing those log ids. Let me think a bit more and come back.